### PR TITLE
Feature/embedded url/177294663

### PR DESCRIPTION
**Pull request by claireolmstead**
Purpose
---------------
Feature

[https://www.pivotaltracker.com/story/show/177294663](https://www.pivotaltracker.com/story/show/177294663)

As a user, when I include a link in the text of a post or reply, it should be clickable.  OnClick, the link is opened in a new tab.

Solution
---------------
When reading the content for posts and replies we want to look for valid URL, replace them with a link tag.

styling:
if a valid URL, it should turn the URL to dark green.

Change summary
---------------
* add a find valid link helper function
* use when displaying content for a post or reply
* add testing

Steps to Verify
----------------
1. Put a link in a caption
2. put a link in a reply
3. click on it
4. see that it takes you to a new page

Additional details / screenshot
----------------
- it was too much of a pain to auto find the URL when you're typing and change the styling, so we just kept that black and only inserted the colored link in the saved post and replies

Screen Shot 2021-11-02 at 12 23 47 PM

### CODE REVIEW

File: src/components/FromTitle.tsx [hunk 1]
```diff
@@ -15,9 +15,9 @@ export const FromTitle = ({
   isHoveringProfile,
   isReply,
 }: FromTitleProps): JSX.Element => {
-  const atHandle = userInfo.handle && "@" + userInfo.handle;
+  const atHandle = userInfo?.handle && "@" + userInfo?.handle;
```
Comment (claireolmstead) [inline src/components/FromTitle.tsx:18]: issues with testing so had to add the "?"s in

File: src/components/test/Feed.test.tsx [hunk 2]
```diff
@@ -139,4 +139,15 @@ describe("Feed", () => {
...
     });
   });
+
+  it("displays link correctly in post", () => {
+    const component = mount(componentWithStore(Feed, store));
+    expect(
+      component.find(".Post__caption").first().prop("dangerouslySetInnerHTML")
+        ?.__html
+    ).toEqual(
+      'test post <a class="messageLink" href=https://www.unfinishedlabs.io/ target="_blank">https://www.unfinishedlabs.io/</a>'
```
Comment (claireolmstead) [inline src/components/test/Feed.test.tsx:149]: linter does not like this.  tried `` as well but prettier changed them to ''. any suggestions?

Comment (wilwade) [reply src/components/test/Feed.test.tsx:149]: @claireolmstead Let's update the .eslintrc.json file rules to this (which will fix it and allow `'I "like" this rule better'`):

OLD: `"quotes": ["error", "double"],`
NEW: `"quotes": ["error", "double", { "avoidEscape": true, "allowTemplateLiterals": true }],`

File: src/utilities/helpers.ts [hunk 3]
```diff
@@ -10,3 +10,10 @@ export const getThumbnailUrl = (url: string): string => {
 
   return url;
 };
+
+export const formatMessage = (message: string): string => {
```
Comment (wilwade) [inline src/utilities/helpers.ts:14]: I think we might want a utility package that does this work for us. Urls can be much more complex. Here are two that might be worth checking on: https://www.npmjs.com/package/react-anchorme or http://markdown-it.github.io/linkify-it/doc/

File: src/components/FromTitle.tsx (hunk 1 above)
Comment (wilwade) [reply src/components/FromTitle.tsx:18]: That would imply that our prop type is wrong. userInfo should exist, but handle might not. I wonder if we can fix the tests to take these optionals back out.

**Review by wilwade — COMMENTED** [review]
Few comments.

**Review by wilwade — APPROVED** [review]
🚂 LTGM!

**Pull request merged**